### PR TITLE
Introduce flow/packet target, local and netflow targets

### PR DESCRIPTION
**Pull request by safchain**

### CODE REVIEW

**Comment by hunchback** [comment]
other that being able to report NetFlowV5 from agent, you also may want to use NetFlowV5 as an alternative for reporting flows from the agents to analyzer, by that also solving limited network bw (by implementing sampling capabilities of NetFlowV5).

**Comment by safchain** [comment]
@hunchback from my point of view, expire/update of flow is already a sampling mechanism.

File: flow/probes/targets/netflow_v5.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+	sysBootMs int64
+	seqNum    uint32
+}
+
+// SendPacket implements the Target interface
+func (nf *NetFlowV5Target) SendPacket(packet gopacket.Packet, bpf *flow.BPF) {
+	nf.table.FeedWithGoPacket(packet, bpf)
+}
+
+func ipStrToUint32(ipAddr string) (uint32, error) {
```
Comment (hunchback) [inline flow/probes/targets/netflow_v5.go]: you may consider placing in common

File: flow/probes/targets/netflow_v5.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+func (nf *NetFlowV5Target) sendFlowBulk(flows []*flow.Flow, now time.Time) {
+	records := new(bytes.Buffer)
+
+	var nFlows uint16
+	for _, f := range flows {
+		if f.Network == nil || (f.Network.Protocol != flow.FlowProtocol_TCP && f.Network.Protocol == flow.FlowProtocol_UDP) {
+			continue
+		}
+
```
Comment (hunchback) [inline flow/probes/targets/netflow_v5.go]: should also process UDP:
```
               if f.Network == nil || (f.Network.Protocol != flow.FlowProtocol_TCP && f.Network.Protocol != flow.FlowProtocol_UDP) {
			continue
		}
```

File: flow/probes/targets/netflow_v5.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+	nf.table.Stop()
+
+	if nf.conn != nil {
+		nf.conn.Close()
+	}
+}
+
+// NewNetFlowV5Target returns a new NetFlow v5 target
+func NewNetFlowV5Target(capture *types.Capture, nodeTID string) (*NetFlowV5Target, error) {
+	u, err := url.Parse("udp://127.0.0.1:6666")
```
Comment (hunchback) [inline flow/probes/targets/netflow_v5.go]: according to standard there are several acceptable port numbers - please make url configurable

Comment (safchain) [reply flow/probes/targets/netflow_v5.go]: sure, this patch is still in draft mode, will put it in the config in a next iteration

**Comment by lebauce** [comment]
@hunchback Quoted from https://stackoverflow.com/questions/41087206/on-github-whats-the-difference-between-reviewer-and-assignee : 

> According to official github documentation, Assignee is a person who is working on specific issues and pull requests. It is sometimes confused as a reviewer. It is actually meant to be used with issues rather than pull request so that when we receive a issue we can assign someone to fix it.

File: common/types.go [hunk 4]
```diff
@@ -470,6 +471,25 @@ func IPV4CIDRToRegex(cidr string) (string, error) {
 	return "^" + regex + `(\/[0-9]?[0-9])?$`, nil
 }
 
+// IsIPv6 returns whether is a IPV6 addresses or not
+func IsIPv6(str string) bool {
```
Comment (lebauce) [inline common/types.go]: return ip != nil && len(ip) == net.IPv6len

Could you please use `IsIPv6` in NormalizeAddrForURL ?

File: flow/probes/targets/erspan.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+		logging.GetLogger().Errorf("Write error erspan packet: %s", err)
+	}
+	ers.buffer.Clear()
+
+	ers.gre.Seq++
+}
+
+// Start start the target
+func (ers *ERSpanTarget) Start() {
+
```
Comment (lebauce) [inline flow/probes/targets/erspan.go]: nit: remove empty line

**Comment by safchain** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by safchain** [comment]
run skydive-functional-tests-backend-elasticsearch

File: common/netns.go [hunk 6]
```diff
@@ -79,3 +79,23 @@ func NewNetNsContext(path string) (*NetNSContext, error) {
 		newns:  newns,
 	}, nil
 }
+
+// ExecInNs call the given function in the context of the given netns
+func ExecInNs(nsPath string, fnc func(origns, newns netns.NsHandle) error) error {
```
Comment (lebauce) [inline common/netns.go]: It seems we don't actually jump into the namespace. So `ExecInNs` is a bit misleading

File: flow/probes/targets/erspan.go [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+// Stop stops the target
+func (ers *ERSpanTarget) Stop() {
+	if ers.fd != 0 {
+		syscall.Close(ers.fd)
+	}
+}
+
+// NewERSpanTarget returns a new ERSpan target
+func NewERSpanTarget(g *graph.Graph, n *graph.Node, capture *types.Capture, nodeTID string) (*ERSpanTarget, error) {
+	fd := C.open_raw_socket(C.uint16_t(syscall.IPPROTO_GRE))
```
Comment (lebauce) [inline flow/probes/targets/erspan.go]: Why not using common.RawSocket ?

Comment (safchain) [reply flow/probes/targets/erspan.go]: with common.RawSocket we can't specify the protocol properly, we can specify the ethertype not the IP protocol

File: common/netns.go (hunk 6 above)
Comment (safchain) [reply common/netns.go]: indeed and not used anymore, going to revert this

File: flow/client/flow_client.go [hunk 8]
```diff
@@ -158,13 +158,16 @@ retry:
...
-// SendFlows sends flows to the server
-func (c *FlowClient) SendFlows(flowArray *flow.FlowArray) {
-	// TODO(safchain) in case of websocket there is no size limitation we should send
-	// bulk directly
-	for _, flow := range flowArray.Flows {
-		if err := c.SendFlow(flow); err != nil {
-			logging.GetLogger().Errorf("Unable to send flow: %s", err)
+// SendFlows implements the flow Sender interface
+func (c *FlowClient) SendFlows(flows []*flow.Flow) {
+	if len(flows) > 0 {
```
Comment (masco-enovance) [inline flow/client/flow_client.go]: since we are using ```range ```, do we need this check?

Comment (safchain) [reply flow/client/flow_client.go]: indeed stupid :)

**Comment by safchain** [comment]
run skydive-scale-tests
run skydive-cdd-overview-tests

**Pull request merged**